materialized_sql_view/model/abstract_materialized_sql_view.py

```diff
@@ -0,0 +1,344 @@
+               rec['view_name'] != self.sql_view_name or
+               rec['state'] in ['nonexistent', 'aborted'] or
+               not pg.is_existed_relation(cr, self.sql_view_name) or
+               not pg.is_existed_relation(cr, self.sql_mat_view_name)):
+                self.drop_materialized_view_if_exist(
+                    cr,
+                    uid,
+                    rec['pg_version'],
+                    view_name=rec['view_name'],
+                    context=context)
+            else:
+                return []
+        return self.create_materialized_view(cr, uid, context=context)
+
+    def change_matview_state(self, cr, uid, method_name, pg_version,
+                             context=None):
+        matview_mdl = self.pool.get('materialized.sql.view')
+        # Make sure object exist or create it
+        values = {
+            'model_name': self._name,
+            'view_name': self.sql_view_name,
+            'matview_name': self.sql_mat_view_name,
+            'sql_definition': self.sql_view_definition,
+            'pg_version': pg_version,
+        }
+        matview_mdl.create_if_not_exist(cr, uid, values, context=context)
+        method = getattr(matview_mdl, method_name)
+        context.update({'values': values})
+        return method(cr, uid, self.sql_mat_view_name, context=context)
+
+    def drop_materialized_view_if_exist(self, cr, uid, pg_version,
+                                        view_name=None, mat_view_name=None,
+                                        context=None):
+        result = []
+        logger.info("Drop Materialized view %r", self.sql_mat_view_name)
+        try:
+            self.before_drop_materialized_view(cr, uid, context=context)
+            pg = PGMaterializedViewManager.getInstance(pg_version)
+            if not view_name:
+                view_name = self.sql_view_name
+            if not mat_view_name:
+                mat_view_name = self.sql_mat_view_name
+            pg.drop_mat_view(cr, view_name, mat_view_name)
+            self.after_drop_materialized_view(cr, uid, context=context)
+        except psycopg2.Error as e:
+            self.report_sql_error(cr, uid, e, pg_version, context=context)
+        else:
+            result = self.change_matview_state(cr, uid, 'after_drop_view',
+                                               pg_version, context=context)
+        return result
+
+    def report_sql_error(self, cr, uid, err, pg_version, context=None):
+        if not context:
+            context = {}
+        context.update({'error_message': err.pgerror})
+        cr.rollback()
+        self.change_matview_state(
+            cr, uid, 'aborted_matview', pg_version, context=context)
+
+    def before_drop_materialized_view(self, cr, uid, context=None):
+        """Method called before drop materialized view and view,
+           Nothing done in abstract method, it's  hook to used in subclass
+        """
```

Comment:
I don't think it's necessary,

Developers don't have to implement this method, it's just hook to let  them to remove index or whatever they want, I guess in most cases we won't implement it